### PR TITLE
issue-558, PostgreSQL user deletion flow was implemented

### LINKED ISSUE #558: Implement PostgreSQL user deletion flow

### PR DESCRIPTION
**Pull request by DoodgeMatvey**
Closes #558

### CODE REVIEW

File: controllers/clusterresources/postgresqluser_controller.go [hunk 1]
```diff
@@ -259,12 +321,72 @@ func (r *PostgreSQLUserReconciler) createUser(
...
+	}
+
+	defaultUsername, defaultPassword, err := getUserCreds(defaultUserSecret)
+	if err != nil {
+		return "", "", "", fmt.Errorf("cannot get default PostgreSQL user credentials, user reference: %v, err: %w", defaultUserSecretNamespacedName, err)
+	}
+
+	clusterName := defaultUserSecret.Labels[models.ControlledByLabel]
+
+	return defaultUsername, defaultPassword, clusterName, nil
```
Comment (OleksiienkoMykyta) [inline controllers/clusterresources/postgresqluser_controller.go:387]: Maybe it would be better to return structure end error instead of multiple arguments

File: controllers/clusterresources/postgresqluser_controller.go [hunk 2]
```diff
@@ -259,12 +321,72 @@ func (r *PostgreSQLUserReconciler) createUser(
...
+) error {
+	defaultUsername, defaultPassword, clusterName, err := r.getDefaultPostgreSQLUserCreds(ctx, defaultUserSecretNamespacedName)
+	if err != nil {
+		return err
+	}
+
+	serviceName := fmt.Sprintf(models.ExposeServiceNameTemplate, clusterName)
+	host := fmt.Sprintf("%s.%s", serviceName, defaultUserSecretNamespacedName.Namespace)
+
+	dbURL := fmt.Sprintf("postgres://%s:%s@%s:%d/%s?target_session_attrs=read-write",
```
Comment (OleksiienkoMykyta) [inline controllers/clusterresources/postgresqluser_controller.go:343]: I think it would be better to move such operations to separate method, because there are multiple places where you use it

Comment (OleksiienkoMykyta) [inline controllers/clusterresources/postgresqluser_controller.go:343]: I think it would be better to move such operations to separate method, because there are multiple places where you use it

**Review by OleksiienkoMykyta — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/postgresqluser_controller.go (hunk 1 above)
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go]: Fixed

File: controllers/clusterresources/postgresqluser_controller.go (hunk 2 above)
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go:346]: Fixed

File: controllers/clusterresources/postgresqluser_controller.go [hunk 3]
```diff
@@ -170,10 +172,86 @@ func (r *PostgreSQLUserReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 			continue
 		}
 
-		// TODO: implement user deletion logic on this event
+		if clusterInfo.Event == models.DeletingEvent {
+			l.Info("Deleting user from a cluster", "cluster ID", clusterID)
+
+			err = r.deleteUser(ctx, newUsername, clusterInfo.DefaultSecretNamespacedName)
```
[the remark this answers is not in the document]
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go:177]: We have default one, and I think to make it more clear it is better to use `new` for naming

File: controllers/clusterresources/postgresqluser_controller.go [hunk 4]
```diff
@@ -170,10 +171,86 @@ func (r *PostgreSQLUserReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+				"Deleting finalizer from the user's secret has been failed. Reason: %v", err)
+			return models.ReconcileRequeue, nil
+		}
+
+		controllerutil.RemoveFinalizer(u, u.GetDeletionFinalizer())
+		err = r.Patch(ctx, u, patch)
+		if err != nil {
+			l.Error(err, "Cannot delete finalizer from the PostgreSQL user resource")
+			r.EventRecorder.Eventf(u, models.Warning, models.PatchFailed,
+				"Deleting finalizer  from the PostgreSQL user resource has been failed. Reason: %v", err)
```
Comment (testisnullus) [inline controllers/clusterresources/postgresqluser_controller.go:246]: Remove an extra space please

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/postgresqluser_controller.go (hunk 4 above)
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go]: Fixed

File: controllers/clusterresources/postgresqluser_controller.go [hunk 5]
```diff
@@ -238,33 +299,101 @@ func (r *PostgreSQLUserReconciler) createUser(
...
+	ctx context.Context,
+	defaultUserSecretNamespacedName clusterresourcesv1beta1.NamespacedName,
+) (*models.Credentials, string, error) {
+	defaultUserSecret := &k8sCore.Secret{}
+
+	namespacedName := types.NamespacedName{
+		Namespace: defaultUserSecretNamespacedName.Namespace,
+		Name:      defaultUserSecretNamespacedName.Name,
+	}
+	err := r.Get(ctx, namespacedName, defaultUserSecret)
```
Comment (testisnullus) [inline controllers/clusterresources/postgresqluser_controller.go:373]: Can we use `defaultUserSecretNamespacedName` instead of `namespacedName`? If yes, let's use it

File: controllers/clusters/postgresql_controller.go [hunk 6]
```diff
@@ -569,6 +569,129 @@ func (r *PostgreSQLReconciler) createUser(
 	return nil
 }
 
+func (r *PostgreSQLReconciler) handleUsersDelete(
+	ctx context.Context,
+	l logr.Logger,
+	c *v1beta1.PostgreSQL,
```
Comment (testisnullus) [inline controllers/clusters/postgresql_controller.go:575]: Why do you refer to PostgreSQL as `c`? Let's use `pg` at least, wdyt?

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/postgresqluser_controller.go (hunk 5 above)
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go:373]: We can't because it is our custom type

File: controllers/clusters/postgresql_controller.go (hunk 6 above)
Comment (DoodgeMatvey) [reply controllers/clusters/postgresql_controller.go]: It is like cluster, but ok, renamed

File: controllers/clusterresources/postgresqluser_controller.go [hunk 7]
```diff
@@ -185,27 +262,11 @@ func (r *PostgreSQLUserReconciler) createUser(
...
-
 		return err
 	}
 
-	defaultUsername, defaultPassword, err := getUserCreds(defaultUserSecret)
-	if err != nil {
-		return fmt.Errorf("cannot get default PostgreSQL user credentials, user reference: %v, err: %w", defaultUserSecretNamespacedName, err)
-	}
-
-	clusterName := defaultUserSecret.Labels[models.ControlledByLabel]
```
Comment (ribaraka) [inline controllers/clusterresources/postgresqluser_controller.go:208]: I think these are unnecessary changes. The old implementation was more straightforward and readable.
Now for example, why does the `r.getDefaultPostgreSQLUserCreds` return a cluster name?

File: controllers/clusterresources/postgresqluser_controller.go [hunk 8]
```diff
@@ -231,40 +292,108 @@ func (r *PostgreSQLUserReconciler) createUser(
...
+		Name:      defaultUserSecretNamespacedName.Name,
+	}
+	err := r.Get(ctx, namespacedName, defaultUserSecret)
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return nil, "", fmt.Errorf("cannot get default PostgreSQL user secret, user reference: %v, err: %w", defaultUserSecretNamespacedName, err)
+		}
+
+		return nil, "", err
+	}
```
Comment (ribaraka) [inline controllers/clusterresources/postgresqluser_controller.go:380]: I think it could be simplified
```suggestion
	if err != nil {
		return nil, "",  fmt.Errorf("cannot get default PostgreSQL user secret, user reference: %v, err: %w", defaultUserSecretNamespacedName, err)
	}
```

File: pkg/models/apiv2.go [hunk 9]
```diff
@@ -169,6 +169,11 @@ type AddonBundle struct {
 	Options *BundleOptions `json:"options"`
 }
 
+type Credentials struct {
+	Username string
+	Password string
+}
+
```
Comment (ribaraka) [inline pkg/models/apiv2.go:176]: This is more of an operator model rather than apiv2.

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/postgresqluser_controller.go (hunk 7 above)
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go:208]: This code is duplicated in some places, so I decided to move it into a separate function. We fetch this creds by defaultUserSecretNamespacedName and I think we need to specify for what cluster they are for because for each cluster they are different

File: controllers/clusterresources/postgresqluser_controller.go (hunk 8 above)
Comment (DoodgeMatvey) [reply controllers/clusterresources/postgresqluser_controller.go:380]: fixed

File: pkg/models/apiv2.go (hunk 9 above)
Comment (DoodgeMatvey) [reply pkg/models/apiv2.go]: fixed

**Review by ribaraka — APPROVED** [review]

**Review by OleksiienkoMykyta — APPROVED** [review]
